core/src/main/java/org/infinispan/commands/write/InvalidateVersionsCommand.java

```diff
@@ -38,39 +45,63 @@ public InvalidateVersionsCommand(ByteString cacheName) {
       super(cacheName);
+      this.topologyId = topologyId;
       this.keys = keys;
       this.topologyIds = topologyIds;
       this.versions = versions;
       this.removed = removed;
    }
 
-   public void init(DataContainer dataContainer, OrderedUpdatesManager orderedUpdatesManager) {
+   public void init(DataContainer dataContainer, OrderedUpdatesManager orderedUpdatesManager,
+                    StateTransferLock stateTransferLock, StateTransferManager stateTransferManager) {
       this.dataContainer = dataContainer;
       this.orderedUpdatesManager = orderedUpdatesManager;
+      this.stateTransferLock = stateTransferLock;
+      this.stateTransferManager = stateTransferManager;
    }
 
    @Override
    public CompletableFuture<Object> invokeAsync() {
-      for (int i = 0; i < keys.length; ++i) {
-         Object key = keys[i];
-         if (key == null) break;
-         SimpleClusteredVersion version = new SimpleClusteredVersion(topologyIds[i], versions[i]);
-         dataContainer.compute(key, (k, oldEntry, factory) -> {
-            if (oldEntry == null) {
-               return null;
-            }
-            SimpleClusteredVersion localVersion = (SimpleClusteredVersion) oldEntry.getMetadata().version();
-            InequalVersionComparisonResult result = localVersion.compareTo(version);
-            if (result == InequalVersionComparisonResult.BEFORE || (removed && result == InequalVersionComparisonResult.EQUAL)) {
-               return null;
-            } else {
-               return oldEntry;
+      if (stateTransferLock != null) {
+         stateTransferLock.acquireSharedTopologyLock();
+      }
+      try {
+         if (topologyId >= 0 && stateTransferManager != null && topologyId != stateTransferManager.getCacheTopology().getTopologyId()) {
+            stateTransferLock.releaseSharedTopologyLock();
+            return CompletableFutures.completedNull();
+         }
+         for (int i = 0; i < keys.length; ++i) {
+            Object key = keys[i];
+            if (key == null) break;
+            SimpleClusteredVersion version = new SimpleClusteredVersion(topologyIds[i], versions[i]);
+            dataContainer.compute(key, (k, oldEntry, factory) -> {
+               if (oldEntry == null) {
+                  return null;
+               }
+               SimpleClusteredVersion localVersion = (SimpleClusteredVersion) oldEntry.getMetadata().version();
+               InequalVersionComparisonResult result = localVersion.compareTo(version);
+               if (result == InequalVersionComparisonResult.BEFORE || (removed && result == InequalVersionComparisonResult.EQUAL)) {
+                  return null;
+               } else {
+                  return oldEntry;
+               }
+            });
+         }
+         return orderedUpdatesManager.invalidate(keys).handle((nil, t) -> {
```

Comment:
I wasn't thinking about persistence, but about not blocking while holding the shared topology lock in general.

`StateTransferLockImpl.ownershipLock` is non-fair, so keeping the read lock for a long time may delay write lock acquisition (IOW topology updates) for a long time. Of course, it may be simpler to make `ownershipLock` fair, but then we'd have to profile it and make sure it's not slower than then non-fair version when no locks are acquired.